_posts/2020-09-29-aws-developer-tools-mule-cicd-solution.md

```diff
@@ -0,0 +1,196 @@
+    - demo-custom-connector/target/*.jar
+    - demo-custom-connector/target/*.zip
+
+cache:
+  paths:
+    - '/root/.m2/**/*'
+```
+
+### env-section
+
+This is where local variables are defined for the Build Project and also the variables that are retrieved from Parameter Store. [See Secrets chapter](#secrets). 
+The value for the varible is the key from Parameter Store and CodeBuild replaces the value with the actual value from Parameter Store during execution.
+
+```yml
+env:
+  variables:
+    key: "value"
+  parameter-store:
+    key: "value-name-for-parameter-store"
+```
+
+### phases-section
+
+Phases-section is where the actual build execution is defined. It is split into two sections: install and build.
+
+In the *install*-section the environment for the build is defined. Here I have selected the JDK to use in the build and set custom Maven profile for the build to use.
+The custom settings files are copied from the sources directory (checked out automatically from Git) in the machine running the build to a special location */root/* so that they are taken into use during execution.
+
+In the *build*-section the commands are listed that actually perform the build. For the `mvn package deploy` step the parameters come from the variables defined in the env-section. In the builds stage, I first build and package a custom connector that is used by the MuleSoft application internally.
+
+### artifacts-section
+
+Here is where you can spcify artifacts from the build process to be stored in the S3 bucket. These can then later on be used in deployments to other Amazon Services or other services that can download from S3.
+
+### cache-section
+
+As we all know, Maven tends to download half of the internet to do a build. To speed up builds we should setup a cache. See the [Cache for Maven repository](#cache-for-maven-repository) section to see how to set it up.
+
+In the *cache*-section of the buildSpec we define the path for the local Maven repository to cache. 
+
+## Secrets
+
+Storing secrets and providing them to the pipeline and build process is something one must almost always do. Documentation guides the developer to use AWS Systems Manager Parameter Store. CodeBuild is then able to access the secrets during runtime. 
+
+After adding the secrects to the Parameter Store, you must add *ssm:GetParameters* permission to the CodeBuild Service Worker role for it to be able to access the secrets. The easiest way to do so is by adding *AmazonSSMReadOnlyAccess* policy for the CodeBuild Service Worker Role.
+
+Maven Secrets are also used to encrypt the credentials needed to access the private Maven repository of MuleSoft and the deployment keys to CloudHub. The process is explained [here](https://docs.mulesoft.com/mule-runtime/3.9/mule-maven-plugin#encrypting-credentials). The actual Maven Master Password needs to be stored somewhere securely. I used AWS Secrects Manager, but any Password manager would do. 
+
+### Cache for Maven repository
```

Comment:
I think this section could be part of the earlier cache section, so it wouldn't then be necessary to reference between the sections.

Reply:
I agree.